clang/lib/AST/ExprConstant.cpp

```diff
@@ -10895,6 +10899,132 @@ bool VectorExprEvaluator::VisitUnaryOperator(const UnaryOperator *E) {
   return Success(APValue(ResultElements.data(), ResultElements.size()), E);
 }
 
+static bool EvaluateVectorOrLValue(APValue &Result, EvalInfo &Info,
+                                   const Expr *E, const QualType &Type) {
+  if (!Evaluate(Result, Info, E))
+    return false;
+
+  if (Result.isLValue()) {
+    // Source of the data is an lvalue; Manually handle the lvalue as if
+    // it was an rvalue to get the current APValue.
+    LValue LValueFound;
+    LValueFound.setFrom(Info.Ctx, Result);
+    if (!handleLValueToRValueConversion(Info, E, Type, LValueFound, Result))
+      return false;
+  }
+
+  return Result.isVector();
+}
+
+static bool handleVectorConversion(EvalInfo &Info, const FPOptions FPO,
+                                   const Expr *E, QualType SourceTy,
+                                   QualType DestTy, APValue const &Original,
+                                   APValue &Result) {
+  if (SourceTy->isIntegerType()) {
+    if (DestTy->isRealFloatingType()) {
+      Result = APValue(APFloat(0.0));
+      return HandleIntToFloatCast(Info, E, FPO, SourceTy, Original.getInt(),
+                                  DestTy, Result.getFloat());
+    }
+    if (DestTy->isIntegerType()) {
+      Result = APValue(
+          HandleIntToIntCast(Info, E, DestTy, SourceTy, Original.getInt()));
+      return true;
+    }
+  } else if (SourceTy->isRealFloatingType()) {
+    if (DestTy->isRealFloatingType()) {
+      Result = Original;
+      return HandleFloatToFloatCast(Info, E, SourceTy, DestTy,
+                                    Result.getFloat());
+    }
+    if (DestTy->isIntegerType()) {
+      Result = APValue(APSInt());
+      return HandleFloatToIntCast(Info, E, SourceTy, Original.getFloat(),
+                                  DestTy, Result.getInt());
+    }
+  }
+  return false;
+}
+
+bool VectorExprEvaluator::VisitConvertVectorExpr(const ConvertVectorExpr *E) {
+  APValue Source;
+  QualType SourceVecType = E->getSrcExpr()->getType();
+  if (!EvaluateVectorOrLValue(Source, Info, E->getSrcExpr(), SourceVecType))
+    return false;
+
+  QualType DestTy = E->getType()->castAs<VectorType>()->getElementType();
+  QualType SourceTy = SourceVecType->castAs<VectorType>()->getElementType();
+
+  const FPOptions FPO = E->getFPFeaturesInEffect(Info.Ctx.getLangOpts());
+
+  auto SourceLen = Source.getVectorLength();
+  SmallVector<APValue, 4> ResultElements;
```

Comment:
I chose 4 as plenty of times I've had to deal with SIMD code it has been either SSE code dealing with floats and doubles and AVX code mainly dealing with doubles, but this certainly is a biased worldview; not all applications deal with doubles and certainly from time to time I've had to code for avx512.

In any case, I think that doubling the smallvector would not be ideal stacksize-wise, and a normal vector with a reserve call would be more than enough to cover all cases. We always know the size and a normal allocation should be fine.

APValue is indeed expensive, but it is also convenient; I think I'm going to stick with it for this PR, and if very long vectors ever get supported this might need to be improved upon. Should I leave a comment explaining that to the reader?

Reply:
Looking at how other code does it, it seems like I also based my decision on the fact that all other vector code uses a size of 4. I think that just for consistency, it is ok to leave it as it is.

Reply:
Yeah, I don't have much of an idea how to evaluate the tradeoffs here: I was hoping you might, and that you'd teach me :)

I agree that it's out of scope for this PR, and that sticking with the 4 for consistency's sake is as good a number as any!

Reply:
304 bytes is definitely inside what I'd be slightly worried to allocate without thinking into the stack, but having looked at it I think it's going to be fine. We're not the only case of smallVectors with APValues, and I'm sure that if it ever becomes a problem the default stack size can be modified (or the code can be refactored, I'm sure that if that ever happens it's not going to be "just" the fault of a smallvector of APValues). I thank you for pointing it out, as I simply did not check!